### PR TITLE
Fix deferred init narrowing #1446

### PR DESCRIPTION
**Pull request by dima-starosud**
### Fix deferred init narrowing https://github.com/hsutter/cppfront/issues/1446 by keeping brace-init in caller context.

- Use construct_from(lambda) so constexpr initializers are not lost across a call boundary, matching immediate initialization behavior.
- Note, change out::construct to use brace-init in all paths to be consistent and to align with deferred init.
- Add regression tests for literals (negative too), constexpr, aggregates.
- Update snapshots affected by the change.

### CODE REVIEW

File: include/cpp2util.h [hunk 1]
```diff
@@ -1488,11 +1488,19 @@ class deferred_init {
...
     constexpr  deferred_init() noexcept       { }
     constexpr ~deferred_init() noexcept       { destroy(); }
     constexpr auto value()     noexcept -> T& { cpp2_default.enforce(init);  return t(); }
 
-    constexpr auto construct(auto&& ...args) -> void { cpp2_default.enforce(!init);  new (&data) T{CPP2_FORWARD(args)...};  init = true; }
+    constexpr auto construct(auto&& ...args) -> void {
+        construct_from([&]() -> T { return T{CPP2_FORWARD(args)...}; });
+    }
+
+    constexpr auto construct_from(auto&& factory) -> void
```
Comment (dima-starosud) [inline include/cpp2util.h:1501]: With the factory, all the values are still on the caller context, which means their constexpr-ness isn't lost during crossing the call boundaries.

File: include/cpp2util.h [hunk 2]
```diff
@@ -1535,10 +1545,16 @@ class out {
...
+        requires std::same_as<decltype(CPP2_FORWARD(factory)()), T>
+    {
         if (has_t || called_construct()) {
-            if constexpr (requires { *t = T(CPP2_FORWARD(args)...); }) {
+            if constexpr (requires { *t = CPP2_FORWARD(factory)(); }) {
                 cpp2_default.enforce( t );
-                *t = T(CPP2_FORWARD(args)...);
+                *t = CPP2_FORWARD(factory)();
```
Comment (dima-starosud) [inline include/cpp2util.h:1557]: Here, instead of parens-init we switched to braces-init to align with deferred init, which is also used in `out` down below.

File: regression-tests/test-results/mixed-autodiff-taylor.cpp [hunk 3]
```diff
@@ -140,44 +140,44 @@ return  { std::move(y0.value()), std::move(y.value()) }; }
       cpp2::impl::deferred_init<double> y0;
       cpp2::impl::deferred_init<taylor> y;
 #line 39 "mixed-autodiff-taylor.cpp2"
-  y0.construct(sqrt(x0));
-  y.construct(CPP2_UFCS(sqrt)(x, x0));
+  y0.construct_from([&]() -> typename CPP2_TYPEOF(y0)::value_type { return typename CPP2_TYPEOF(y0)::value_type{sqrt(x0)}; });
+  y.construct_from([&]() -> typename CPP2_TYPEOF(y)::value_type { return typename CPP2_TYPEOF(y)::value_type{CPP2_UFCS(sqrt)(x, x0)}; });
```
Comment (dima-starosud) [inline regression-tests/test-results/mixed-autodiff-taylor.cpp:144]: We could introduce a macro for this. If we decide to, I can bring it in this PR.

File: regression-tests/test-results/clang-19-c++20/pure2-bugfix-for-deferred-init-narrowing-matrix-error.cpp.output [hunk 4]
```diff
@@ -0,0 +1,57 @@
+pure2-bugfix-for-deferred-init-narrowing-matrix-error.cpp2:19:125: error: non-constant-expression cannot be narrowed from type 'cpp2::impl::in<cpp2::i16>' (aka 'const short') to 'signed char' in initializer list [-Wc++11-narrowing]
```
Comment (dima-starosud) [inline regression-tests/test-results/clang-19-c++20/pure2-bugfix-for-deferred-init-narrowing-matrix-error.cpp.output:1]: This tests errors, i.e. failing cases, when we do want the compilation to fail.
